crmprtd/download_cache_process.py

```diff
@@ -0,0 +1,371 @@
+    if network_name == "ec":
+        net_args = f"-p {province.lower()} -F {frequency}".split()
+    if network_name in (
+        "bc_env_snow",
+        "bc_forestry",
+        "bc_tran",
+        "dfo_ccg_lighthouse",
+        "nt_forestry",
+        "nt_water",
+        "yt_gov",
+        "yt_water",
+        "yt_avalanche",
+        "yt_firewx",
+    ):
+        ts = to_utc(time).strftime("%Y/%m/%d %H:00:00")
+        net_args = ["-d", f'"{ts}"']
+    if network_name == "moti":
+        net_args = f"-u https://prdoas5.apps.th.gov.bc.ca/saw-data/sawr7110 --auth_fname ~/.rtd_auth.yaml --auth_key={network_name}".split()
+    if network_name == "wamr":
+        net_args = []
+    if network_name == "wmb":
+        net_args = f"--auth_fname ~/.rtd_auth.yaml --auth_key={network_name}".split()
+
+    assert net_args is not None
+    return common_args + net_args
+
+
+def alias_to_networks(network_alias: str):
+    return network_aliases[network_alias]
+
+
+def dispatch_network(connection_string: str = None, **kwargs) -> None:
+    """
+    Dispatch a single network to the download-and-process pipeline.
+
+    :param connection_string: Database connection string for "process" step.
+    :param kwargs: Remaining args, passed through to various subfunctions. Note that
+        network name is one of these args.
+    :return: None. Side effect: Download and process network specified in args.
+    """
+    network_name = kwargs["network_name"]
+    check_network_name(network_name)
+
+    if network_name == "ec":
+        for province in ("BC", "YT"):
+            download_and_process(
+                network_name=network_name,
+                log_args=log_args(**kwargs, province=province),
+                download_args=download_args(**kwargs, province=province),
+                cache_filename=cache_filename(**kwargs, province=province),
+                connection_string=connection_string,
+            )
+    elif network_name in (
+        "bc_env_snow",
+        "bc_forestry",
+        "bc_tran",
+        "dfo_ccg_lighthouse",
+        "nt_forestry",
+        "nt_water",
+        "yt_gov",
+        "yt_water",
+        "yt_avalanche",
+        "yt_firewx",
+    ):
+        an_hour_ago = datetime.datetime.now() - datetime.timedelta(hours=1)
+        download_and_process(
+            network_name=network_name,
+            log_args=log_args(**kwargs),
+            download_args=download_args(**kwargs, time=an_hour_ago),
+            cache_filename=cache_filename(**kwargs, timestamp=an_hour_ago),
+            connection_string=connection_string,
+        )
+    else:
+        now = datetime.datetime.now()
+        download_and_process(
+            network_name=network_name,
+            log_args=log_args(**kwargs),
+            download_args=download_args(**kwargs, time=now),
+            cache_filename=cache_filename(**kwargs, timestamp=now),
+            connection_string=connection_string,
+        )
```

Comment:
Is moti also a special case? `infill.infill` has this:

```python
    if "moti" in networks:

        # Query all existing stations
        stations = get_moti_stations(connection_string)

        # MoTI has an insane config where each user has a specific set
        # of stations associated with it. PCIC wants *all* the
        # stations which is too many for one request, so we have two
        # users with half of the stations. This way of requesting the
        # data will get many "skips", but at least we'll get all of
        # the data.
        for auth_key in ("moti", "moti2"):
            # Divide range into 6 day intervals
            for interval_start, interval_end in zip(
                weekly_ranges[:-1], weekly_ranges[1:]
            ):
                for station in stations:
                     ...
```

but the execution script simply had this:

```bash
echo "crmptd_download -N ${net} -u https://prdoas5.apps.th.gov.bc.ca/saw-data/sawr7110 --auth_fname ~/.rtd_auth.yaml --auth_key=${net} -L ~/logging.yaml --log_filename ${log_file} | tee ${cache_file} |\
 crmprtd_process -c ${db} -N ${net} -L ~/logging.yaml --log_filename ${log_file}" | batch
```

Why are these so different?

Reply:
Technically *every* network is a "special case". That's why this software exists... to allow us to download data according to the specific ways that each network expose it to us.

In the case of MoTI, there are two endpoints:

1. The default endpoint which gives the last hour of data from all stations
2. A "historic" endpoint where you can provide a station id and time range (no greater than 7 days) and get historic data. The infill script is specifically tailored to this use case for obvious reasons.

Reply:
Because of this, running a historic query is much more heavyweight than running regular hourly updates from the default endpoint (and its preferable to use the latter). Can you think of a good way to encode this into the script behavior and/or the invocation options?

Reply:
Point taken re. "special cases". 

The info on MoTI is very helpful, thanks. I think it should be feasible to allow for both historical infill and hourly updates. That said, that's what `infill_all` still does. I didn't change its behaviour. Options:
1. Leave d-c-p and infill_all separate, and use them separately. No effort beyond what's already been done; a bit clunky but it works.
2. Merge infill_all functionality into d-c-p. More effort; possibly a nicer interface.

Preference?

Reply:
I've created a separate [issue](https://github.com/pacificclimate/crmprtd/issues/136) for this. The question is more general than just for MoTI, although I think that is the gnarliest one.